stdlib/LinearAlgebra/src/uniformscaling.jl

```diff
@@ -284,6 +284,24 @@ end
     mul!(C, A, J.λ, alpha, beta)
 @inline mul!(C::AbstractVecOrMat, J::UniformScaling, B::AbstractVecOrMat, alpha::Number, beta::Number) =
     mul!(C, J.λ, B, alpha, beta)
+
+function mul!(out::AbstractMatrix{T}, a::Number, B::UniformScaling, α::Number, β::Number) where {T}
+    checksquare(out)
+    if iszero(β)  # zero contribution of the out matrix
+        fill!(out, zero(T))
+    elseif !isone(β)
+        rmul!(out, β)
+    end
+    s = convert(T, a*B.λ*α)
+    if !iszero(s)
+        @inbounds for i in diagind(out)
+            out[i] += s
+        end
+    end
+    return out
+end
+@inline mul!(out::AbstractMatrix, A::UniformScaling, b::Number, α::Number, β::Number)=
+    mul!(out, b, A, α, β)
```

Comment:
This is dangerous, because the eltypes can be `Quaternion` or something non-commutative, and then changing the order of multiplication gives wrong results.

Reply:
Wow. It is hard to write good julia code. Is it documented somewhere that `mul!` should respect non-commutativity? Or that the type `Number` does not necessarily mean commutativity? Anyway, another approach committed.

Reply:
Well, it depends on how you read the docstring. It says something like computes `C <- A*B*α + C*β`, and when you have non-commutative numbers, then the ordering is relevant, and you'd get a different result if you computed `C <- α*A*B + β*C`, for instance. Or in other words: is it documented that `Number` should mean commutativity? 😛

Reply:
Yes. It would be good to know what can be assumed about a `Number`. To extend the reasoning ab absurdum: In many places (eg in many `mul!`-implementations) it is assumed that `Number`s are commutative with respect to addition.  How do I know that `alpha+beta=beta+alpha`? Without that assumption it is difficult to tune efficiency. My point is that it would have been nice to have this explicit somewhere. Then I would not have to bother a julia member with a long rant. 😉

Reply:
Good point. I think addition is generally assumed to be commutative, but multiplication not. But there is an everyday example for that: matrix multiplication!

Don't worry about bothering people. We're all here to make the world a better place and Julia an even better language. 🤣